py/src/braintrust/contrib/temporal/__init__.py

```diff
@@ -0,0 +1,483 @@
+            type="task",
+            parent=parent_span_context or None,
+            metadata={
+                "temporal.activity_type": info.activity_type,
+                "temporal.activity_id": info.activity_id,
+                "temporal.workflow_id": info.workflow_id,
+                "temporal.workflow_run_id": info.workflow_run_id,
+            },
+        )
+        span.set_current()
+
+        try:
+            result = await super().execute_activity(input)
+            return result
+        except Exception as e:
+            span.log(error=str(e))
+            raise
+        finally:
+            span.unset_current()
+            span.end()
+
+
+class BraintrustWorkflowInboundInterceptor(temporalio.worker.WorkflowInboundInterceptor):
+    """Workflow interceptor that creates workflow spans and propagates context to activities.
+
+    This interceptor creates a span for the workflow execution using sandbox_unrestricted
+    to bypass Temporal's sandbox restrictions. The workflow span is the parent for all
+    activities and child workflows executed within it.
+    """
+
+    def __init__(self, next: temporalio.worker.WorkflowInboundInterceptor) -> None:
+        super().__init__(next)
+        self.payload_converter = temporalio.converter.PayloadConverter.default
+        self._parent_span_context: Optional[Dict[str, Any]] = None
+
+    def init(self, outbound: temporalio.worker.WorkflowOutboundInterceptor) -> None:
+        super().init(_BraintrustWorkflowOutboundInterceptor(outbound, self))
+
+    async def execute_workflow(
+        self, input: temporalio.worker.ExecuteWorkflowInput
+    ) -> Any:
+        # Extract parent span context from workflow headers (set by client)
+        parent_span_context = None
+        if _HEADER_KEY in input.headers:
+            header_payload = input.headers.get(_HEADER_KEY)
+            if header_payload:
+                payloads = self.payload_converter.from_payloads([header_payload])
+                if payloads:
+                    parent_span_context = payloads[0]
+
+        # Store parent span context for activities (will be overwritten if we create a workflow span)
+        self._parent_span_context = parent_span_context
+
+        # Create a span for the workflow execution using sandbox_unrestricted
+        # to bypass the sandbox restrictions on logger state access
+        span = None
+        if not temporalio.workflow.unsafe.is_replaying():
+            with temporalio.workflow.unsafe.sandbox_unrestricted():
+                # Get logger via extern function (supports test logger parameter)
+                get_logger = temporalio.workflow.extern_functions()["__braintrust_get_logger"]
+                logger = get_logger()
+
+                if logger:
+                    info = temporalio.workflow.info()
+                    span = logger.start_span(
+                        name=f"temporal.workflow.{info.workflow_type}",
+                        type="task",
```

Comment:
i don't have enough context, but is there anything that could be a function instead of a task?

Reply:
i don't know what the difference is?